cloud/scope/powervs_cluster.go

```diff
@@ -882,37 +901,98 @@ func (s *PowerVSClusterScope) ReconcileNetwork() (bool, error) {
-	if s.IBMPowerVSCluster.Spec.Network.ID != nil {
-		s.V(3).Info("Checking if PowerVS network exists in IBM Cloud with ID", "networkID", *s.IBMPowerVSCluster.Spec.Network.ID)
-		network, err := s.IBMPowerVSClient.GetNetworkByID(*s.IBMPowerVSCluster.Spec.Network.ID)
+// checkNetwork checks if network exists in cloud with given network's ID or name or dhcp server's ID or name mentioned in spec.
+// Priority order: Network ID - Network name - DHCP server ID - DHCP server name - Network with cluster name.
+func (s *PowerVSClusterScope) checkNetwork() (*string, *string, error) {
+	checkNetworkByID := func(networkID string) (*string, error) {
+		s.V(3).Info("Checking if PowerVS network exists in IBM Cloud with ID", "networkID", networkID)
+		network, err := s.IBMPowerVSClient.GetNetworkByID(networkID)
 		if err != nil {
 			return nil, err
 		}
 		return network.NetworkID, nil
 	}
 
+	if s.IBMPowerVSCluster.Spec.Network.ID != nil {
+		networkID, err := checkNetworkByID(*s.IBMPowerVSCluster.Spec.Network.ID)
+		if err != nil {
+			return nil, nil, err
+		}
+		dhcpServerID, err := s.verifyDHCPServerRespectToNetwork(*networkID)
+		if err != nil {
+			return nil, nil, err
+		}
+		return dhcpServerID, networkID, nil
+	}
+
+	checkNetworkByName := func(networkName string) (*string, *string, error) {
+		s.V(3).Info("Checking if PowerVS network exists in IBM Cloud with name", "name", networkName)
+		network, err := s.IBMPowerVSClient.GetNetworkByName(networkName)
+		if err != nil {
+			return nil, nil, err
+		}
+		if network == nil || network.NetworkID == nil {
+			s.V(3).Info("Unable to find PowerVS network in IBM Cloud", "network", s.IBMPowerVSCluster.Spec.Network)
+			return nil, nil, nil
+		}
+		dhcpServerID, err := s.verifyDHCPServerRespectToNetwork(*network.NetworkID)
+		if err != nil {
+			return nil, nil, err
+		}
+		return dhcpServerID, network.NetworkID, nil
+	}
+
+	if s.IBMPowerVSCluster.Spec.Network.Name != nil {
+		dhcpServerID, networkID, err := checkNetworkByName(*s.IBMPowerVSCluster.Spec.Network.Name)
+		if err != nil {
+			return nil, nil, err
+		}
+		return dhcpServerID, networkID, nil
+	}
+
+	if s.DHCPServer() != nil && s.DHCPServer().ID != nil {
+		dhcpServer, err := s.IBMPowerVSClient.GetDHCPServer(*s.DHCPServer().ID)
+		if err != nil {
+			return nil, nil, err
+		}
+		networkID, err := checkNetworkByID(*dhcpServer.Network.ID)
+		if err != nil {
+			return nil, nil, err
+		}
+		return dhcpServer.ID, networkID, nil
+	}
+
 	// if the user has provided the already existing dhcp server name then there might exist network name
 	// with format DHCPSERVER<DHCPServer.Name>_Private , try fetching that
 	var networkName string
 	if s.DHCPServer() != nil && s.DHCPServer().Name != nil {
-		networkName = fmt.Sprintf("DHCPSERVER%s_Private", *s.DHCPServer().Name)
+		networkName = dhcpNetworkName(*s.DHCPServer().Name)
 	} else {
 		networkName = *s.GetServiceName(infrav1beta2.ResourceTypeNetwork)
```

Comment:
is this required? as we are already checking this condition above

Reply:
Can you please give me the ref? I think this is the only place where I am checking with cluster name.

Reply:
Thanks for clarifying, I understand it is required to preserve the order.

Reply:
Sure, addressed your other comments. ptal.